### ISSUE TITLE
Query : not all Nortek wave parameters are stored (compared with RDI waves)

### ISSUE THREAD

**Issue by sspagnol** [aodn/imos-toolbox#468]
From browsing awacParse.m and readAWACWaveAscii.m it would appear that not all wave parameters are stored in the resultant `sample_data` structure; for example `waveData.PeakPeriod` which I would have would be mapped to IMOS variable WPPE (cf workhorseParse.m and waveData.param.Tp -> WPPE).

So just wanted to check if this is delibrate design decision (i.e. the Nortek wave parameters aren't equivalent to IMOS wave variable definitions) or just one of those things thas has fallen through the cracks.

If it's something that has been missed can it be looked at (can attempt if you wish). As still haven't quite got a handle on why the 31/15 column versions of wap file, readAWACWaveAscii.m will still have to handle that with the if statement. And modification to awacParse.m will be to extend the `vars` wave parameter mappings, and to also to test for the existence of a particular field before adding it to `sample_data` structure.  I don't know if there is a standard IMOS variable to store SpectraType but that could be handy.

Have attached the Storm help page text on Standard Wave Parameters (text only) for reference

[storm_wave_parameters.txt](https://github.com/aodn/imos-toolbox/files/1237925/storm_wave_parameters.txt)

I would also like to make the the code handling the addition of wave data to `sample_data` (basically anything after

```matlab
%
% if wave data files are present, read them in
%
waveData = readAWACWaveAscii(filename);

% no wave data, no problem
if isempty(waveData), return; end
```
be moved into a helper function for code reuse. I have an aquadopp profiler that can measure waves (limited to 1Hz sampling frequency I think, and limited to pressure method), so would like to also modify aquadoppProfilerParse.m by adding similar read and test for waveData etc.

**Comment by ggalibert** [comment]
Several requests here:

1. From browsing awacParse.m and readAWACWaveAscii.m it would appear that not all wave parameters are stored in the resultant sample_data structure; for example waveData.PeakPeriod which I would have would be mapped to IMOS variable WPPE (cf workhorseParse.m and waveData.param.Tp -> WPPE).

This is a mistake. Any other field you want to see added to sample_data apart from PeakPeriod and SpectraType?

2. I would also like to make the the code handling the addition of wave data to ```sample_data``` be moved into a helper function for code reuse. I have an aquadopp profiler that can measure waves (limited to 1Hz sampling frequency I think, and limited to pressure method), so would like to also modify aquadoppProfilerParse.m by adding similar read and test for waveData etc.

Will you use readAWACWaveAscii to read data from your Aquadopp?

**Comment by ggalibert** [comment]
spectraType is actually already added to the NetCDF file with IMOS parameter code SPCT.

**Comment by ggalibert** [comment]
The following fields are not added to the NetCDF file for now:
```
waveData.Battery
waveData.MinPressure
waveData.MaxPressure
waveData.PeakPeriod
waveData.PeakDirection
waveData.UnidirectivityIndex
```
Agree we should add PeakPeriod and PeakDirection. Any other one you think would be worth adding too?

**Comment by sspagnol** [comment]
Missed the SPCT in the list :)

Would have thought from the Nortek WAP column definitions something like

| waveData column | Description | IMOS variable |
| --- | --- | ---|
|   8  | Significant height (Hm0)        (m) | WWSH |
|   9  | Mean 1/3 height (H3)            (m) | WHTH |
| 10  | Mean 1/10 height (H10)        (m) | WHTE |
| 11  | Maximum height (Hmax)        (m) | WMXH (AST only) |
| 12  | Mean Height (Hmean)            (m) | WMSH (AST only) |
| 13  | Mean  period (Tm02)              (s) | WPFM see Nortek wave parameter definition |
| 14  | Peak period (Tp)                      (s) | WPPE |
| 15  | Mean zerocrossing period (Tz) (s) | WPMH |
| 16  | Mean 1/3 Period (T3)              (s) | WPTH |
| 17  | Mean 1/10 Period (T10)          (s) | WPTE |
| 18  | Maximum Period (Tmax)          (s) | WMPP |
| 19  | Peak direction (DirTp)              (deg) | WPDI |
| 20  | Directional spread (SprTp)       (deg) | ?, possibly no CF equivalent |
| 21  | Mean direction (Mdir)             (deg) | ?, possibly no CF equivalent |

But would need a waves expert to check these mappings (especially WPFM mapping).

And yes, aquadoppProfilerParse.m would use readAWACWaveAscii, test if return waveData array, if not empty call new helper function to create additional sample_data structure. Didn't know if you wanted it as one big change or say three changes (add new vars, make new helper function, modify aquadoppProfilerParse)

**Comment by ggalibert** [comment]
Yes, ideally 3 distinct commits would be nice. Will look at your mapping tomorrow.

**Comment by sspagnol** [comment]
Actually think "Mean Period (Tm02) (s)" should be WPSM. Just to note that there is a difference between Storm wave parameter help file definition

```
Mean Period (Tm02)                        

This is the average period for all the waves in the burst and it is calculated from the energy 
spectrum according to the first and second moment of the energy spectrum: Tm02 = sum(M0/M02) The value is reported in seconds.
```

and CF definition

```
The wind wave directional spectrum can be written as a five dimensional function S(t,x,y,f,theta) where t is time, x and y are horizontal coordinates (such as longitude and latitude), f is frequency and theta is direction. S can be integrated over direction, thus S1= integral(S dtheta). Frequency moments, M(n) of S1 can then be calculated as follows: M(n) = integral(S1 f^n df), where f^n is f to the power of n. The second wave period, T(m2), is calculated as the square root of the ratio M(0)/M(2).
```

But responses in Nortek forum [here](http://www.nortek-as.com/en/knowledge-center/forum/waves/20001901?b_start=0#778128530) and [here](http://www.nortek-as.com/en/knowledge-center/forum/waves/20001873?b_start=0#103163061) and Nortek's [technical note](http://www.nortek-as.com/lib/technical-notes/puv-wave-measurement) would suggest it's just a typo in Storm help file.

**Comment by ggalibert** [comment]
I am adding to your previous table the current mapping performed by the toolbox in order to show clearer the differences with what you suggest:

| waveData column | Nortek description | Nortek definition | current IMOS param | CF standard name | To update to? | CF standard name |
| --- | --- | --- | --- | --- | --- | --- |
|   8  | Significant height (Hm0) (m) | This is the classic estimate sometimes referred to as Hs. It is calculated from the energy spectrum, Hmo = 4sqrt(sum(M0)) | WSSH | sea surface wave spectral significant height **not CF** | WWSH | sea surface wind wave significant height
|   9  | Mean 1/3 height (H3) (m) | This is the mean of the 1/3 largest waves in a record. It is a time series based estimate. Typically this value is 5% larger than Hmo, yet variations can be greater or smaller. Note: AST only | none yet | | WHTH | sea surface wave significant height |
| 10  | Mean 1/10 height (H10) (m) | This is the mean of the 1/10 largest waves in a record. It is a time series based estimate when AST is available. When AST is not available, then this estimate may simply be presented as a linear extrapolation of Hmo, whereby H10 = 1.27Hm0. Note: AST only | none yet | | WHTE | sea surface wave mean height of highest tenth |
| 11  | Maximum height (Hmax) (m) | This is the largest wave in a record. It is a time series based estimate when AST is available. When AST is not available, then this estimate may simply be presented as a linear extrapolation of Hmo, whereby Hmax = 1.67Hm0. Note: AST only | none yet | | WMXH (AST only) | sea surface wave maximum height |
| 12  | Mean Height (Hmean) (m) | This is the mean value of all waves in a record. It is a time series based estimate when AST is available. When AST is not available, then this estimate is marked as an invalid value and not displayed | none yet | | WMSH (AST only) | sea surface wave mean height |
| 13  | Mean  period (Tm02) (s) | This is the average period for all the waves in the burst and it is calculated from the energy spectrum according to the first and second moment of the energy spectrum: Tm02 = sqrt(M0/M02) The value is reported in seconds | none yet | | WPSM | sea surface wave mean period from variance spectral density second frequency moment |
| 14  | Peak period (Tp) (s) | This is the period of the waves corresponding to the peak frequency for the wave spectrum. The value is reported in seconds | WPPE | sea surface wave period at variance spectral density maximum | WPPE | sea surface wave period at variance spectral density maximum |
| 15  | Mean zerocrossing period (Tz) (s) | This is the mean period calculated from the zero-crossing technique. It is calculated as the mean of all the periods in the wave burst. The value is reported in seconds | VAVT | sea surface wave zero upcrossing period | WPMH | sea surface wave mean period |
| 16  | Mean 1/3 Period (T3) (s) | This is the mean period associated with the 1/3 largest waves (H3) in a record, where the period is calculated from the zero-crossing technique. The value is reported in seconds | none yet | | WPTH | sea surface wave significant period |
| 17  | Mean 1/10 Period (T10) (s) | This is the mean period associated with the 1/10 largest waves (H10) in a record, where the period is calculated from the zero-crossing technique. The value is reported in seconds | none yet | | WPTE | sea surface wave mean period of highest tenth |
| 18  | Maximum Period (Tmax) (s) | This is the mean period associated with the largest wave (Hmax) in a record, where the period is calculated from the zero-crossing technique. The value is reported in seconds | none yet | | WMPP | sea surface wave maximum period |
| 19  | Peak direction (DirTp) (deg) | This is the direction of the wave corresponding to the peak period. The direction is reported as “from” and is reported in degrees | WPDI | sea surface wave from direction at variance spectral density maximum | WPDI | sea surface wave from direction at variance spectral density maximum |
| 20  | Directional spread (SprTp) (deg) | The directional spread is a measure of the directional variance. The estimate is calculated for the peak frequency. The value is reported in degrees | SSDS | sea surface wave directional spread **not CF** | ?, possibly no CF equivalent | |
| 21  | Mean direction (Mdir) (deg) | This value is a weighted average of all the directions in the wave spectrum. It is weighted according to the energy at each frequency. The direction is reported as “from” and is reported in degrees | VDIR | sea surface wave from direction **not CF** | ?, possibly no CF equivalent | |

I agree with all your suggestions except for column 8. Could you please explain?

**Comment by sspagnol** [comment]
Just a typo on my part . And just to clarify, its WSSH and not VAVH because of the method used to calculate it?

**Comment by ggalibert** [comment]
Yes, it's WSSH and not VAVH (because of method used)

**Comment by ggalibert** [comment]
BTW, I'm already working on the implementation to fix this issue...

**Comment by sspagnol** [comment]
Cool.

**Comment by ggalibert** [comment]
This has been fixed in v2.5.32 via the following commits:
https://github.com/aodn/imos-toolbox/commit/ad46cbd9be38dcdc0d27d022a5afc18f27ff9b38
https://github.com/aodn/imos-toolbox/commit/1f73834a27afb0347a14471772087f3963576d26
